core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -130,30 +137,57 @@ KOKKOS_INLINE_FUNCTION
     std::enable_if_t<std::is_same<T, unsigned long long>::value, T>
         cast_from_half(half_t);
 
-class alignas(2) half_t {
+template <class FLOAT_TYPE>
+class alignas(FLOAT_TYPE) floating_point_wrapper {
  public:
-  using impl_type = Kokkos::Impl::half_impl_t::type;
+  using impl_type = FLOAT_TYPE;
 
  private:
   impl_type val;
+  using volatile_type = typename std::conditional<
+      sizeof(FLOAT_TYPE) == 2, uint16_t,
+      typename std::conditional<
+          sizeof(FLOAT_TYPE) == 4, uint32_t,
+          typename std::conditional<sizeof(FLOAT_TYPE) == 8, uint64_t,
+                                    void>::type>::type>::type;
+  static_assert(!std::is_same<volatile_type, void>::value,
+                "Invalid FLOAT_TYPE");
+
+  // Casting wrappers for supporting multiple impl types
+  static KOKKOS_INLINE_FUNCTION impl_type float2impl_type(float x) {
+    if (std::is_same<impl_type, Kokkos::Impl::half_impl_t::type>::value)
+      return __float2half(x);
+  }
+  static KOKKOS_INLINE_FUNCTION float impl_type2float(impl_type x) {
```

Comment:
`impl_` prefix for a member function typically means "private" implementation in Kokkos